### PR TITLE
Add a Datasets API notebook

### PR DESCRIPTION
**Pull request by AtreyeeS**
This PR adds an API notebook for `Datasets`
I think there is sufficient room for improvement, so suggestions much welcome @registerrier @QRemy @luca-giunti @adonath @LauraOlivera @bkhelifi 
I think the `FluxPoints` can be explained in a dedicated `estimators` notebook, and we can link to that.

### CODE REVIEW

File: docs/tutorials/api/datasets.ipynb [hunk 1]
```diff
@@ -0,0 +1,877 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Datasets - Reduced data, IRFs, models \n",
+    "\n",
+    "## Introduction\n",
+    "\n",
+    "`gammapy.datasets` are a crucial part of the gammapy API. `datasets` constitute `DL4` data - binned counts, IRFs, models and the associated likelihoods. `Datasets` from the end product of the `makers` stage, see [makers notebook](makers.ipynb), and are passed on to the `Fit` or estimator classes for modelling and fitting purposes.\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:11]: About "`Datasets` from the end product of the `makers` stage", is it not rather:
"`Datasets` are the products of the Data Reduction stage, with the use of `makers`, [...]"

File: docs/tutorials/api/datasets.ipynb [hunk 2]
```diff
@@ -0,0 +1,877 @@
...
+    "\n",
+    "There are three types of `Dataset` available\n",
+    "- `MapDataset`: Binned counts and IRFs on a `WCSGeom` spatial geom and an energy axis; used for 3D analysis, supports cash likelihood\n",
+    "- `SpectrumDataset`: Binned counts and IRFs on a `RegionGeom` and energy axis; used for 1D spectral analysis; supports cash likelihood\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:15]: it is rather wstat instead of cash, no?

File: docs/tutorials/api/datasets.ipynb [hunk 3]
```diff
@@ -0,0 +1,877 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Datasets - Reduced data, IRFs, models \n",
+    "\n",
+    "## Introduction\n",
+    "\n",
+    "`gammapy.datasets` are a crucial part of the gammapy API. `datasets` constitute `DL4` data - binned counts, IRFs, models and the associated likelihoods. `Datasets` from the end product of the `makers` stage, see [makers notebook](makers.ipynb), and are passed on to the `Fit` or estimator classes for modelling and fitting purposes.\n",
+    "\n",
+    "There are three types of `Dataset` available\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:13]: Should we add a web link towards the statistics page?

File: docs/tutorials/api/datasets.ipynb [hunk 4]
```diff
@@ -0,0 +1,877 @@
...
+    "from gammapy.estimators import FluxPoints"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## MapDataset\n",
+    "\n",
+    "The counts, exposure, background, masks, and IRF maps are bundled together in a data structure named `MapDataset`. While the `counts`, and `background` maps are binned in reconstructed energy must have the same geometry, the IRF maps can have a different spatial (coarsely binned and larger) geometry and spectral range (binned in true energies). It is usually recommened that the true energy bin should be larger and more finely sampled and the reco energy bin.\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:59]: something is missing in this sentence:
"While the `counts`, and `background` maps are binned in reconstructed energy must [...]"

Comment (bkhelifi) [reply docs/tutorials/api/datasets.ipynb:59]: Also "recommened' -> "recommended" ;)

File: docs/tutorials/api/datasets.ipynb [hunk 5]
```diff
@@ -0,0 +1,877 @@
...
+    "    width=5 * u.deg,\n",
+    "    binsz=0.05 * u.deg,\n",
+    "    frame=\"icrs\",\n",
+    ")\n",
+    "\n",
+    "energy_axis_true = MapAxis.from_energy_bounds(\n",
+    "    0.1, 100, nbin=11, name=\"energy_true\", unit=\"TeV\", per_decade=True\n",
+    ")\n",
+    "\n",
+    "rad_axis = MapAxis.from_bounds(0, 5, nbin=50, unit=\"deg\", name=\"rad\")\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:93]: the rad_axis is really optionnal. Should we add it in this tuto?

File: docs/tutorials/api/datasets.ipynb [hunk 6]
```diff
@@ -0,0 +1,877 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Using masks\n",
+    "\n",
+    "There are two masks that can be set on a `Dataset`, `mask_safe` and `mask_fit`. \n",
+    "\n",
+    "- The `mask_safe` is computed during the data reduction process according to the specified selection cuts, and should not be changed by the user.\n",
+    "- During modelling and fitting, the user might want to additionally ignore some parts of a reduced dataset, e.g. to restrict the fit to a specific energy range or to ignore parts of the region of interest. This should be done bu applying the `mask_fit`. To details of applying masks, please refer to [Masks-for-fitting](mask_maps.ipynb#Masks-for-fitting:-mask_fit)\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:234]: "This should be done bu" -> "This should be done by"

File: docs/tutorials/api/datasets.ipynb [hunk 7]
```diff
@@ -0,0 +1,877 @@
...
+    "execution": {
+     "iopub.execute_input": "2021-08-31T15:44:59.271898Z",
+     "iopub.status.busy": "2021-08-31T15:44:59.271573Z",
+     "iopub.status.idle": "2021-08-31T15:45:03.682747Z",
+     "shell.execute_reply": "2021-08-31T15:45:03.682380Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# To apply a mask fit - in enegy and space\n",
```
Comment (bkhelifi) [inline docs/tutorials/api/datasets.ipynb:269]: enegy -> energy

File: docs/tutorials/api/datasets.ipynb (hunk 1 above)
Comment (AtreyeeS) [reply docs/tutorials/api/datasets.ipynb:11]: Thanks @bkhelifi ! I meant `form` not `from` :-/

File: docs/tutorials/api/datasets.ipynb (hunk 3 above)
Comment (AtreyeeS) [reply docs/tutorials/api/datasets.ipynb:13]: Ok

File: docs/tutorials/api/datasets.ipynb [hunk 8]
```diff
@@ -0,0 +1,877 @@
+{
```
Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Technically we have seen that there might be cases where a <code>MapDataset</code>  could have a <code>RegionGeom</code>.  </p><p>I would also stress that <code>SpectrumDataset</code>  has no <code>psf</code>  associated.</p><p> </p><p>Maybe an array with datasets as rows and data type (i.e. counts of flux), reduced IRFs present, geometry and stat could be helpful and graphical? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Maybe a subtitle to explain you are creating an empty dataset first</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>saved to disk</p><p> </p><p>I would also mention that this saves the various <code>Maps</code>  as various HDUs  of a single FITS file. And that the <code>Maps</code>  are currently stored according to the gadf specifications for skymaps.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Maybe somewhere the importance and the usage of the dataset names could be stressed.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>This an aparte but I think we need a <code>MapDataset.peek</code>  method as well.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Shouldn't we explore other contents of a Dataset?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>I think serialization should be discussed briefly here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Maybe mention that they can be used for catalog objects and point towards notebooks where they are used/fitted.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (registerrier) [inline docs/tutorials/api/datasets.ipynb:1]: <p>I think you could discuss joint vs stacked analysis and present <code>Datasets</code>  as the tool to handle the former.</p><p> </p><p>I would present some of the functionalities here: I/O of course but also stack_reduce and maybe some others.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

**Review by registerrier — COMMENTED** [review]
Thanks @Atreyee . This is a very useful addition.

I have left some comments inline. I think it would be good to provide a tabular presentation of the various datasets and their contents.

What about the part on stacking that is present in the API doc. Should you link to it? Move some part in?

File: docs/tutorials/api/datasets.ipynb [hunk 9]
```diff
@@ -0,0 +1,1163 @@
...
```
Comment (adonath) [inline docs/tutorials/api/datasets.ipynb:1]: <p>I think the overview table is very useful, but maybe move it to <code>docs/datasets.rst</code>?  And put a link here instead?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (adonath) [inline docs/tutorials/api/datasets.ipynb:1]: <p>I'd say this is one API level  too "low". In the datasets tutorial we should only cover the attributes and methods of the <code>Dataset</code>  objects. The explanation of the <code>PSFMap</code>  does not belong here. Also the <code>PSFKernelMap</code>  object does no exist.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (adonath) [inline docs/tutorials/api/datasets.ipynb:1]: <p>Similar here, I think this documentation rather belongs to the <code>EDispKernelMap</code>  and <code>gammapy.irf</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

**Comment by adonath** [comment]
Thanks a lot @AtreyeeS! I think this is a very good start for the Dataset API notebook. I have left a few inline comments to address. My only remaining comment is that some functionality is missing such as downsampling / resampling etc., but this can be added any time later in follow up PRs.

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS . This looks good.

Does the table render well with Sphinx?

**Pull request merged**